### ISSUE TITLE
ReflectionTypeLoadException in ReflectionDiscovery when trying to get the DefinedTypes from an assembly that cannot be loaded.

### ISSUE THREAD

**Issue by roald-di** [petabridge/NBench#161]
Hi,

I am trying to run a benchmark from a unit test (xunit), but I'm getting a ReflectionTypeLoadException.
The exception is thrown from ReflectionDiscovery.LoadAllTypeConfigurators when it calls Assembly.DefinedTypes [here](https://github.com/petabridge/NBench/blob/126e1d9c832f57039a2f598ab47003216f22c728/src/NBench/Sdk/Compiler/ReflectionDiscovery.Configurators.cs#L82).

This only happens with the visual studio test runner (I'm using vs 2017 rc).

In this case the problem is that the test-runner assembly is not loadable, but in general there is no guarantee that a certain assembly will load.

The 'right way' to to get the types from an assembly, as explained in this [blog post](http://haacked.com/archive/2012/07/23/get-all-types-in-an-assembly.aspx/), seems to be the following:

```
public static IEnumerable<Type> GetLoadableTypes(this Assembly assembly)
{
    if (assembly == null) throw new ArgumentNullException("assembly");
    try
    {
        return assembly.GetTypes();
    }
    catch (ReflectionTypeLoadException e)
    {
        return e.Types.Where(t => t != null);
    }
}
```

I tried building the project with this change and it solved the problem for me.

This problem also impacts the Pro.NBench.xUnit project as seen in this issue which provides more details: Pro-Coded/Pro.NBench.xUnit#2

There is one more place where 'DefinedTypes' is [used](https://github.com/petabridge/NBench/blob/126e1d9c832f57039a2f598ab47003216f22c728/src/NBench/Sdk/Compiler/ReflectionDiscovery.cs#L143), I haven't had an issue with it yet but it might be worth it to fix that as well.

**Comment by mcquiggd** [comment]
I have revisited the issue on the Pro.NBench.XUnit project, and the issue is still present with the following configuration:

Visual Studio 2015 Enterprise (all update versions), Visual Studio 2017 Enterprise 15.2 and 15.3 (preview)
NBench 1.0.4
Any and all combinations of MSTest Framework and MSTest Adaptor, including 1.2.0-beta.
Builds for X86 or x64

In all cases, the Resharper Test Runner executes without difficulty; the issue, as mentioned by the original poster, only affects Visual Studio Test Explorer.

**Comment by mcquiggd** [comment]
I have subsequently tested this with Visual Studio 2017, using [Pro.NBench.Xunit](https://github.com/Pro-Coded/Pro.NBench.xUnit), which targets .NetStandard 1.6. All seems to function correctly.

**Comment by Aaronontheweb** [comment]
I'm getting this myself when unit testing with XUnit, after making some changes to our test suite. Going to fix this as part of the v1.5. release I'm cooking up now.

**Comment by Aaronontheweb** [comment]
This is now fixed as part of https://github.com/petabridge/NBench/pull/229
